discovery_engine_core/bindings/src/types/string.rs

```diff
@@ -0,0 +1,132 @@
+// Copyright 2022 Xayn AG
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, version 3.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+//! FFI functions for handling `String`
+
+use std::ptr;
+
+/// Creates a rust `String` with given capacity at given memory address.
+///
+/// The `String` is created at given address, not it's content.
+///
+/// # Safety
+///
+/// It must be valid to write a `String` instance to given pointer,
+/// the pointer is expected to point to uninitialized memory.
+#[no_mangle]
+pub unsafe extern "C" fn init_string_at(place: *mut String, capacity: usize) -> *mut u8 {
+    let mut s = String::with_capacity(capacity);
+    let data_ptr = s.as_mut_ptr();
+    unsafe {
+        ptr::write(place, s);
+    }
+    data_ptr
+}
+
+/// Sets the length of the rust `String` at given memory address.
+///
+/// Use this after you write to the strings data buffer behind
+/// `len` (but before it's end indicated by `capacity`) to make
+/// the newly written string data available to rust.
+///
+/// # Safety
+///
+/// The pointer must point to a valid `String` instance.
+#[no_mangle]
+pub unsafe extern "C" fn set_string_len(string_place: *mut String, len: usize) {
+    unsafe {
+        (*string_place).as_mut_vec().set_len(len);
+    }
+}
+
+/// Returns the length of an rust `String` at given memory address.
+///
+/// # Safety
+///
+/// The pointer must point to a valid `String` instance.
+#[no_mangle]
+pub unsafe extern "C" fn get_string_len(string: *mut String) -> usize {
+    unsafe { (*string).len() }
+}
+
+/// Returns a pointer to the underlying buffer of the given rust string.
+///
+/// You can write valid utf8 to the buffer up to a length of the strings
+/// `capacity`, after which you can use `set_string_len` to make the written
+/// data available to rust.
+///
+/// # Safety
+///
+/// The pointer must point to a valid `String` instance.
+#[no_mangle]
+pub unsafe extern "C" fn get_string_buffer(string: *mut String) -> *mut u8 {
+    unsafe { &mut *string }.as_mut_ptr()
+}
+
+/// Alloc an uninitialized `Box<String>`, mainly used for testing.
+#[cfg(feature = "additional-ffi-methods")]
+#[no_mangle]
+pub extern "C" fn alloc_uninitialized_string_box() -> *mut String {
+    use super::boxed::alloc_uninitialized_box;
+
+    alloc_uninitialized_box()
```

Comment:
are there limitations which force us to import this locally? otherwise we could just import it at the top of the module as usual or we could call it here with the full path. same for `drop_string_box()`.
```suggestion
    super::boxed::alloc_uninitialized_box()
```

Reply:
It's only imported once and only if the feature which feature gates the function is enabled.

Hence rust-analyzer auto places it in the function.

I will switch it to prefix.